src/context.rs

```diff
@@ -0,0 +1,244 @@
+use std::{
+    any::{Any, TypeId},
+    cmp::Ordering,
+    collections::{BinaryHeap, HashMap, HashSet, VecDeque},
+};
+
+use derivative::Derivative;
+
+pub trait Plugin: Any {
+    type DataContainer;
+
+    fn get_data_container() -> Self::DataContainer;
+}
+
+#[macro_export]
+macro_rules! define_plugin {
+    ($plugin:ident, $data_container:ty, $default: expr) => {
+        struct $plugin {}
+
+        impl $crate::context::Plugin for $plugin {
+            type DataContainer = $data_container;
+
+            fn get_data_container() -> Self::DataContainer {
+                $default
+            }
+        }
+    };
+}
+pub use define_plugin;
+
+pub struct PlanId {
+    pub id: u64,
```

Comment:
Maybe this should not have pub visibility? People using Context should only work with PlanIds that are returned and not construct them directly.